### ISSUE TITLE
Router.generate should support routes registered with sibling routers

### ISSUE THREAD

**Issue by bryanrsmith** [aurelia/router#89]
Currently `router.generate()` will search for named routes registered with the current router, or an ancestor router. Consider a StackOverflow-style site with a `Questions` VM that configures a child router with 'list', 'search', and 'details' routes, and a `Users` VM that configures a child router with 'list', 'search', and 'profile' routes. The question details page will need to link to the user profile page of the asker, and the user profile page will need to link to the question details for questions asked and answered by the user. Currently this can only be done by manually constructing an absolute URL using string concatenation. It would be great if `router.generate()` could handle routes registered in other branches of the router tree.

**Comment by bryanrsmith** [comment]
What if we supported route namespaces, so that from the `Questions` VM, I could `router.generate('users.details', { id: '123' })`. This would activate the 'users' module if it hasn't been configured yet, and then look for the 'details' route under that child router.

**Comment by bryanrsmith** [comment]
After #90 we should be able to support generating for eagerly configured routers.

**Comment by larsfjerm** [comment]
I've been waiting for a fix for this in a long time now. Would this be possible when the beta comes out? I would say it's a crucial functionality.

**Comment by bryanrsmith** [comment]
I agree. I really want this too. I don't think it will be ready before
beta, though. It requires a pretty big internal change that I haven't had
time for. Hoping to get to it soon after beta.

**Comment by larsfjerm** [comment]
Nice to hear. You've all done a great job so far. Looking forward for the rest!

**Comment by migajek** [comment]
Any news on this?
I think this is a must-have for the complex routing solution.
I'm from Angular background and I really like their's de-facto standard routing solution, ui router.

It's hierarchy is defined through namespacing, following/extending a bit @bryanrsmith 's example:

```
Questions 
   Questions.list
   Questions.search
   Questions.details
Users
   Users.list
   Users.search
   Users.details
      Users.details.bio
      Users.details.experience
```

the feature I love there is relational navigation, very intuitive.

Assuming ``Users.details`` is our active (current) route:
``^.list`` == ``Users.list``
``.experience`` == ``Users.details.experience``

I believe the example is self-explanatory ;)

This hierarchical namespacing approach has several major upsides; 
it doesn't require the user to type the whole "path" each time, just the relative part - both when configuring and referencing the route
it makes it easier to maintain complex apps (CRUD especially, where one have tens of "list", "detail", "create", "update" routes)

I'd love to see something like that in Aurelia.

**Comment by peisenmann** [comment]
:+1: Proposed solutions on this page make sense, and I'd love to see them in a coming version. This seems like it could be achieved in beta without any pending changes from jspm, babel, or any of the other dependencies needed for the final stable 1.0 of Aurelia

**Comment by Jaans** [comment]
I too would be keen for a implementation that achieves @migajek 's sample. What are you guys doing to workaround this at the moment? 

One hack-ish thought I'm having is to Inject the singleton viewmodel with the other child routes Router instance and then to perform a code base navigation (instead of href). Other ideas?

**Comment by glen-84** [comment]
> What if we supported route namespaces, so that from the Questions VM, I could router.generate('users.details', { id: '123' }). This would activate the 'users' module if it hasn't been configured yet, and then look for the 'details' route under that child router.

1. I'm already using periods to name my routes, so if there are plans to use periods for route namespacing, they should be made reserved characters.

2. I don't like the idea of activating other modules to access routing data – if you need to generate a full navigation structure, it would cause almost all modules to be activated/downloaded. I would prefer something like #90 where you configure all the routes up front.

**Comment by MaartenThomassen** [comment]
Any progress on this? I agree with @larsfjerm  that this is basic functionality. Especially since the RedirectToRoute class supports parameters and the Redirect class does not. The only way i can now redirect to a child sibbling is to manually format the entire url.

**Comment by jkehler** [comment]
I'm going to chime in here and say this feature would be very very very nice to have. Too bad there's been no progress on it for over a year now.

**Comment by EisenbergEffect** [comment]
Most of our core team has full-time day jobs just like most other open source projects. If something doesn't happen it's because our team has other higher priorities. If you want to see a feature fast-tracked the best way to see that happened is to get involved. It doesn't mean you have to implement the whole thing. There are lots of ways to help:

* Write some unit tests to validate the the functionality you want
* Analyze the codebase and come up with a strategy to implement the feature
* Collect a set of concrete use cases that should be supported
* Actually do the implementation
* etc

You can submit PRs for any of these things and they would be welcomed.

**Comment by heruan** [comment]
This would be great! I would stick with the existing `generate` function, with an algorithm like this:

Being in `/start/from/here`, call `router.generate('get/me/there')`; then `generate` would:

1. split `'get/me/there'` in `[ 'get', 'me', 'there ]` (these are route names);
2. look backwards in the hierarchy for a route named `'get'`;
3. if found (could also be the current router itself), from the router which has the `'get'` route get its view-model (i.e. `moduleId`) from the current container and check for a `configureRouter` function (otherwise throw);
4. if exists, create a child router and configure it with the view-model's `configureRouter`, the call the `generate` function of the child router with the remaining parts of the split path, i.e. `childRouter.generate('me/there')`.

I'm not taking into account route parameters, which could be necessary in any step of the generation. For example, should both `'get'` and `'me'` routes need an `id` parameter, it gets tricky.

**Comment by migajek** [comment]
@EisenbergEffect I believe this is kind of a design/architecture decision, as it'd cause the modules to be (kind of) eager-loaded.
Therefore personally I wouldn't risk taking an effort of implementing proof of concept unless I know the opinion of some core members, as it might be totally wrong approach.

**Comment by EisenbergEffect** [comment]
Understood. What we really need to do is expand the route configuration api so that child routers can be declared up front, making generation possible, but without eager loading.

**Comment by migajek** [comment]
does that mean the declaration of child module's router would be (physically) separate from the actual module, making them not-so-much-self-contained, or did I missed the concept?

**Comment by EisenbergEffect** [comment]
At present, I don't know of a way to achieve both. So, you would have to pick one model or the other. We're are open to ideas.

**Comment by hypnoboutique** [comment]
I don't know much about much but I wrote a gulp task that builds a routes.js file containing all my app routes and chucked it into the build series.

If this sounds like a plausible solution (or starting point) for others, I'm happy to share. Just don't judge me... it was my first foray into Gulp tasks.

Also I'm interested to hear why this might **not** be a good solution, if you think so!

**Comment by JoshMcCullough** [comment]
Maybe I'm late to the party here. If I have `public` and `admin` routes in the root router, e.g.:

```
{ route: "public", name: "public", moduleId: "public/index" },
{ route: "admin", name: "admin", moduleId: "admin/index" }
```

Each of those index viewmodels registers more routes within that area. Now, if I'm at `admin/users/edit` and want to link to my `user-profile` route within my `public` area, it seems I should be able to this (again, from within `admin` area):

```
<a route-href="route: public/user-profile; ...">View profile</a>
```

1. The router would look for a route _local_ named `public/user-profile`, but wouldn't find one.
2. It would defer to the parent router who would look for a local (now root) route named `public/user-profile`, but it wouldn't find one either.
3. The (root) router would then consider the route as path segments, splitting on `/`. It would find my `public` route and load that resource/module.
4. The (root) router would then defer to the `public` router, passing the remainder of the route name, `user-profile`.
5. The (public) router would find this route as a perfect match, and finally return the proper route.

**Comment by RichiCoder1** [comment]
As an alternate to @migajek 's idea, maybe allow passing and array to the relevant parameter? So that `router.generate(["user", "profile"], { user: 12345 })` would check the `user` route, pull the appropriate model & configure (if necessary), then check the `profile` route and generate from there?

I'd possibly be interested in taking a swing at this.

**Comment by tdamir** [comment]
@RichiCoder1 The solution I have for this is inspired by http://www.jeremyg.net/entry/create-a-menu-with-child-routes-using-aurelia. After the app starts, we resolve all the modules in advance based on the initial router. After that we can use this service for URL generation in similar manner you suggested. Here is the code: https://gist.github.com/tdamir/34861b86c015e3fb5c0e25d477b11bbb

**Comment by RichiCoder1** [comment]
@tdamir Neat! I like it :). Will need to explore how that'd best be integrated into the router itself, as well as how to configure eager loading (which is important for things like code splitting and other optimization techniques).

**Comment by Alexander-Taran** [comment]
@EisenbergEffect added to vNext router use cases list
Can be closed
